web/src/main/java/se/inera/intyg/webcert/web/event/CertificateEventServiceImpl.java

```diff
@@ -0,0 +1,332 @@
+
+        if (certificate.getStatus() == UtkastStatus.DRAFT_COMPLETE) {
+            CertificateEvent savedEvent = save(certificateId, WEBCERT_USER, EventCode.KFSIGN, certificate.getKlartForSigneringDatum());
+            events.add(savedEvent);
+        } else if (certificate.getStatus() == UtkastStatus.SIGNED) {
+            Signatur signature = certificate.getSignatur();
+            CertificateEvent savedEvent = save(certificateId, signature.getSigneradAv(), EventCode.SIGNAT, signature.getSigneringsDatum(),
+                "Certificate type: " + certificate.getIntygsTyp());
+            events.add(savedEvent);
+            if (certificate.getAterkalladDatum() != null) {
+                savedEvent = save(certificateId, WEBCERT_USER, EventCode.MAKULERAT, certificate.getAterkalladDatum());
+                events.add(savedEvent);
+            }
+        }
+        return events;
+    }
+
+    private Optional<CertificateEvent> createEventFromSigned(String certificateId, Utlatande utlatande) {
+        GrundData grundData = utlatande.getGrundData();
+
+        if (grundData.getSigneringsdatum() != null) {
+            String signedBy = grundData.getSkapadAv().getPersonId();
+            CertificateEvent savedEvent = save(certificateId, signedBy, EventCode.SIGNAT, grundData.getSigneringsdatum(),
+                utlatande.getTyp());
+            return Optional.of(savedEvent);
+        }
+        return Optional.empty();
+    }
+
+
+    private Optional<CertificateEvent> createEventFromSent(Utkast certificate) {
+        if (certificate.getSkickadTillMottagareDatum() != null) {
+            CertificateEvent savedEvent = save(certificate.getIntygsId(), WEBCERT_USER, EventCode.SKICKAT,
+                certificate.getSkickadTillMottagareDatum(),
+                "Recipient: " + certificate.getSkickadTillMottagare());
+            return Optional.of(savedEvent);
+        }
+        return Optional.empty();
+    }
+
+    private List<CertificateEvent> createEventsFromStatuses(String certificateId, IntygContentHolder certificate) {
+        List<CertificateEvent> events = new ArrayList<>();
+
+        if (!CollectionUtils.isEmpty(certificate.getStatuses())) {
+            for (Status status : certificate.getStatuses()) {
+                if (status.getType().name().equals("SENT") && status.getTimestamp() != null) {
+                    String message = "Recipient: " + status.getTarget();
+                    CertificateEvent savedEvent = save(certificateId, UNKNOWN_USER, EventCode.SKICKAT, status.getTimestamp(), message);
+                    events.add(savedEvent);
+                }
+            }
+        }
+        return events;
+    }
+
+    private List<CertificateEvent> createEventsFromArende(String certificateId) {
+        return createEventsFromArende(certificateId, null);
+    }
+
+    public List<CertificateEvent> createEventsFromArende(String certificateId, LocalDateTime afterTimestamp) {
```

Comment:
Should this be public?

Reply:
To be able to test it(?)

Reply:
OK! Understand! Do you think the test gets too complicated by managing it through getCertificateEvents(..)?  

If you have to test it directly consider making it protected or default (package), so it is not exposed more than necessary.

Reply:
Could also add a @VisibleForTesting.